src/Toml/BiMap.hs

```diff
@@ -138,158 +170,169 @@ _Just = prism Just id
+    eitherInteger n
+      | n < 0     = Left $ ArbitraryError "Value is below zero."
+      | otherwise = Right (fromIntegral n)
 
 -- | 'Natural' bimap for 'AnyValue'. Usually used with 'natural' combinator.
-_Natural :: BiMap Natural AnyValue
+_Natural :: TomlBiMap Natural AnyValue
 _Natural = _NaturalInteger >>> _Integer
 
 -- | Helper bimap for 'Integer' and integral, bounded values.
-_BoundedInteger :: (Integral a, Bounded a) => BiMap a Integer
-_BoundedInteger = BiMap (Just . toInteger) maybeBounded
+_BoundedInteger :: (Integral a, Bounded a) => TomlBiMap a Integer
+_BoundedInteger = BiMap (Right . toInteger) eitherBounded
   where
-    maybeBounded :: forall a. (Integral a, Bounded a) => Integer -> Maybe a
-    maybeBounded n
-      | n < toInteger (minBound :: a) = Nothing
-      | n > toInteger (maxBound :: a) = Nothing
-      | otherwise                     = Just (fromIntegral n)
+    eitherBounded :: forall a. (Integral a, Bounded a) => Integer -> Either TomlBiMapError a
+    eitherBounded n
+      | n < toInteger (minBound :: a) = Left $ ArbitraryError "Value is less minBound"
+      | n > toInteger (maxBound :: a) = Left $ ArbitraryError "Value is more maxBound"
+      | otherwise                     = Right (fromIntegral n)
 
 -- | 'Word' bimap for 'AnyValue'. Usually used with 'word' combinator.
-_Word :: BiMap Word AnyValue
+_Word :: TomlBiMap Word AnyValue
 _Word = _BoundedInteger >>> _Integer
 
 -- | 'Int' bimap for 'AnyValue'. Usually used with 'int' combinator.
-_Int :: BiMap Int AnyValue
+_Int :: TomlBiMap Int AnyValue
 _Int = _BoundedInteger >>> _Integer
 
 -- | 'Float' bimap for 'AnyValue'. Usually used with 'float' combinator.
-_Float :: BiMap Float AnyValue
+_Float :: TomlBiMap Float AnyValue
 _Float = iso realToFrac realToFrac >>> _Double
 
 -- | Helper bimap for 'Text' and strict 'ByteString'
-_ByteStringText :: BiMap ByteString Text
-_ByteStringText = prism T.encodeUtf8 maybeText
-  where
-    maybeText :: ByteString -> Maybe Text
-    maybeText = either (const Nothing) Just . T.decodeUtf8'
+_ByteStringText :: TomlBiMap ByteString Text
+_ByteStringText = prism T.encodeUtf8 (eitherToBiEither . T.decodeUtf8')
 
 -- | 'ByteString' bimap for 'AnyValue'. Usually used with 'byteString' combinator.
-_ByteString:: BiMap ByteString AnyValue
+_ByteString:: TomlBiMap ByteString AnyValue
 _ByteString = _ByteStringText >>> _Text
 
 -- | Helper bimap for 'Text' and lazy 'ByteString'
-_LByteStringText :: BiMap BL.ByteString Text
-_LByteStringText = prism (TL.encodeUtf8 . TL.fromStrict) maybeText
+_LByteStringText :: TomlBiMap BL.ByteString Text
+_LByteStringText = prism (TL.encodeUtf8 . TL.fromStrict) eitherText
   where
-    maybeText :: BL.ByteString -> Maybe Text
-    maybeText = either (const Nothing) (Just . TL.toStrict) . TL.decodeUtf8'
+    eitherText :: BL.ByteString -> Either TomlBiMapError Text
+    eitherText = either (const $ Left $ ArbitraryError "Decoding failed") (Right . TL.toStrict) . TL.decodeUtf8'
```

Comment:
I think it's better to preserve the exact error of `decodeUtf8'` function here to tell what exactly went wrong.